src/ZODB/MappingStorage.py

```diff
@@ -269,9 +267,9 @@ def tpc_begin(self, transaction, tid=None):
         if transaction is self._transaction:
             raise ZODB.POSException.StorageTransactionError(
                 "Duplicate tpc_begin calls for same transaction")
-        self._lock_release()
+        self._lock.release()
         self._commit_lock.acquire()
-        self._lock_acquire()
+        self._lock.acquire()
```

Comment:
Maybe the `utils.RLock` needs to grow a second context manager helper, to handle a scoped release-then-acquire.